balancer/proposals/balancer.md

```diff
@@ -0,0 +1,181 @@
+   Name string
+   // ScaleTargetRef is a reference that points to a target resource to balance.
+   // The target needs to expose the Scale subresource.
+   ScaleTargetRef hpa.CrossVersionObjectReference
+   // MinReplicas is the minimum number of replicas inside of this target.
+   // Balancer will set at least this amount on the target, even if the total
+   // desired number of replicas for Balancer is lower.
+   // +optional
+   MinReplicas *int32
+   // MaxReplicas is the maximum number of replicas inside of this target.
+   // Balancer will set at most this amount on the target, even if the total
+   // desired number of replicas for the Balancer is higher.
+   // +optional
+   MaxReplicas *int32
+}
+ 
+// BalancerPolicyName is the name of the balancer Policy.
+type BalancerPolicyName string
+const (
+   PriorityPolicyName     BalancerPolicyName = "priority"
+   ProportionalPolicyName BalancerPolicyName = "proportional"
+)
+ 
+// BalancerPolicy defines Balancer policy for replica distribution.
+type BalancerPolicy struct {
+   // PolicyName decides how to balance replicas across the targets.
+   // Depending on the name one of the fields Priorities or Proportions must be set.
+   PolicyName BalancerPolicyName
+   // Priorities contains detailed specification of how to balance when balancer
+   // policy name is set to Priority.
+   // +optional
+   Priorities *PriorityPolicy
+   // Proportions contains detailed specification of how to balance when
+   // balancer policy name is set to Proportional.
+   // +optional
+   Proportions *ProportionalPolicy
+   // Fallback contains specification of how to recognize and what to do if some
+   // replicas fail to start in one or more targets. No fallback happens if not-set.
+   // +optional
+   Fallback *FallbackPolicy
+}
+ 
+// PriorityPolicy contains details for Priority-based policy for Balancer.
+type PriorityPolicy struct {
+   // TargetOrder is the priority-based list of Balancer targets names. The first target
+   // on the list gets the replicas until its maxReplicas is reached (or replicas
+   // fail to start). Then the replicas go to the second target and so on. MinReplicas
+   // is guaranteed to be fulfilled, irrespective of the order, presence on the
+   // list, and/or total Balancer's replica count.
+   TargetOrder []string
+}
+ 
+// ProportionalPolicy contains details for Proportion-based policy for Balancer.
+type ProportionalPolicy struct {
+   // TargetProportions is a map from Balancer targets names to rates. Replicas are
+   // distributed so that the max difference between the current replica share
+   // and the desired replica share is minimized. Once a target reaches maxReplicas
+   // it is removed from the calculations and replicas are distributed with
+   // the updated proportions. MinReplicas is guaranteed for a target, irrespective
+   // of the total Balancer's replica count, proportions or the presence in the map.
+   TargetProportions map[string]int32
```

Comment:
There's one thing we might want to note here. We probably should use a method that doesn't result in [Alabama paradox](https://en.wikipedia.org/wiki/Apportionment_paradox#Alabama_paradox).

For example (mapped from Wikipedia link above) we could have:

Deployment name | Proportion | With Balancer Size 10 | With Balancer Size 11
--- | --- | --- | ---
A | 6 |  4 ( rounded down from 4.286) |  5 (rounded up from 4.714)
B | 6 |  4 ( rounded down from 4.286) | 5 (rounded up from 4.714)
C | 2 |  2 (rounded up from 1.429) | 1 (rounded down from 1.571)

So if Balancer size oscillates between 10 and 11 scale up would result in starting 2 pods (and terminating 1). Scale up would result in terminating 2 pods (and starting 1). Which would be bad (and I guess might be very hard to debug).

Ways to do proportional pod assignment are described in the article. 

Also in the article are some other paradoxes but I don't think we should care about them:
- There's a theorem saying we can't avoid *all* of them,
- Other "paradoxes" will happen when `TargetProportion` changes so I think  they're much smaller problem. Something unintuitive happening rarely, on user action sounds less bad than something unintuitive happening potentially frequently, on automated action.

Reply:
I think [Huntington-Hill](https://en.wikipedia.org/wiki/Huntington%E2%80%93Hill_method) would work good:

1) It works nicely with minimums (first step is assigning minimum allowed instances to each deployment)
2) It doesn't have Alabama paradox (so when increasing number of instances in Balancer we won't decrease number of instances in any of its deployments)
3) It's easy to implement:
  1. Assign minimum instances to each deployment
  2. Assign priority to each deployment, put them all in a priority queue
  3. While sum of deployment sizes < Balancer size:
    - Increase desired size of the deployment with the highest priority by 1
    - Update priority of the deployment
4) It should be efficient enough. With N deployments and Balancer size M  it takes something like O((M + N)log(N)) (if we use a heap we need to build heap (N operations) then do M inserts & deleted, each O(log(N)))

Reply:
Very interesting, thanks for sharing the information. We could use the method proposed :).